ocaml/typing/typemod.ml

```diff
@@ -782,6 +782,7 @@ let merge_constraint initial_env loc sg lid constr =
             by the caller. So what we do with the scope doesn't really matter. But
             making it local makes it unlikely that we will ever use the result of
             this function unfreshened without issue. *)
+          let sg = Mtype.expand_to initial_env sg !real_ids in
```

Comment:
Hmm, that's surprising since I (think I) only spotted this issue because of test failures. This is necessary because in, say,
`S with module M.N := X` the destructive substitution is implemented by actually substituting `X` for `M.N` everywhere in `S`. But if somewhere in `S`, we encounter, `Mty_strengthen(T,M)`, then that doesn't work - we need to expand the strengthening until we've materialised all `M.N` in `T` and only then can we substitute.
I'll add a test and a comment.

Reply:
Added a comment and a test.